src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -4066,6 +4066,22 @@ static QVariant fcnStringToArray( const QVariantList &values, const QgsExpressio
   return array;
 }
 
+static QVariant fcnArrayLambda( const QVariantList &values, const QgsExpressionContext *ctx, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QVariantList array = QgsExpressionUtils::getListValue( values.at( 0 ), parent );
+  QString expString = QgsExpressionUtils::getStringValue( values.at( 1 ), parent );
+  QVariantList output;
+  for ( QVariantList::const_iterator it = array.constBegin(); it != array.constEnd(); ++it )
+  {
+    QString expressionString( expString );
+    expressionString.replace( "@", ( *it ).toString() );
```

Comment:
I'd rather see a proper variable used here, inserted into the expression context. So something like

```
  QgsExpressionContextScope* varScope = new QgsExpressionContextScope();
  varScope->addVariable( StaticVariable( QStringLiteral("element"), QVariant(), true );
  context.appendScope( varScope );

  QgsExpression expression( expString  );
  expression.prepare( ctx );

  for ( QVariantList::const_iterator it = array.constBegin(); it != array.constEnd(); ++it )
  {
    varScope->setVariable( QStringLiteral("element"), *it );
    output << expression.evaluate( ctx );
  }

  delete ctx.popScope();
```

This has the benefit that the expression is prepared once only, so there's a massive performance improvement. It also avoids the fragile string replace of '@' - which will break any expression which uses an '@' string literal, or other variables.